File: python/Programmers/보석쇼핑.py
def solution(gems):
    answer = []
    # 모든 보석의 종류
    type_cnt = len(set(gems))
    min_term = len(gems)

    # 투포인터
    start = 0
    end = 0
    buy_type_cnt = dict()
    buy_type_cnt[gems[start]] = 1

    while start<= end:
        if len(buy_type_cnt) == type_cnt:
            if min_term > end-start:
                min_term = end-start
                answer = [start+1, end+1]
            else:
                if buy_type_cnt[gems[start]] == 1:
                    del buy_type_cnt[gems[start]]
                else:
                    buy_type_cnt[gems[start]] -= 1
                start += 1
        elif len(buy_type_cnt) < type_cnt:
            end += 1
            if end == len(gems):
                break
            if gems[end] not in buy_type_cnt:
                buy_type_cnt[gems[end]] = 1
            else:
                buy_type_cnt[gems[end]] += 1
    return answer

File: kotlin/src/Programmers/보석쇼핑.kt
fun solution(gems: Array<String>): IntArray {
    var answer = intArrayOf()

    var typelist = gems.toSet()
    var min_term = gems.size

    var start:Int = 0
    var end:Int = 0

    var dict = HashMap<String,Int>()
    dict[gems[0]]=1

    while(start<=end){
        if(dict.count() == typelist.size){
            if(min_term > end-start){
                min_term = end-start
                answer = intArrayOf(start+1,end+1)

            }else{
                if(dict.containsKey(gems[start])) {
                    var num:Int = dict[gems[start]]!!
                    dict[gems[start]] =  num-1
                }
                if (dict[gems[start]] == 0){
                    dict.remove(gems[start])
                }
                start += 1
            }
        }else if(dict.count() < typelist.size){
            end += 1
            if(end == gems.size) {break}
            if(gems[end] in dict){
                if(dict.containsKey(gems[end])) {
                    var num:Int = dict[gems[end]]!!
                    dict[gems[end]] = num + 1
                }

            }else{
                dict[gems[end]]=1
            }
        }
    }
    return answer
}